### PR TITLE
Apply lua gumbo fixes

### PR DESCRIPTION
**Pull request by stevecheckoway**
This large PR pulls in all of @craigbarnes's fixes to gumbo by reverting all my recent patches to the error reporting, applying all of his commits, and then re-implementing my few commits.

I tried pulling the pieces in piecemeal and making individual PRs for them, but that approach didn't work. Once error reporting is fixed, we could move to a submodule and more easily track lua-gumbo's fixes to gumbo. (Or we could just stay on top of any changes and pull those in ourselves which shouldn't be too hard as long as our copies don't diverge too much.)

### CODE REVIEW

**Comment by stevecheckoway** [comment]
I think we should investigate pulling the gumbo tests in before merging this. Last time I looked, the tests didn't build, but it's worth looking into in more detail.

**Comment by stevecheckoway** [comment]
@craigbarnes Awesome! I really appreciate all your help (and hard work on gumbo!)

**Comment by stevecheckoway** [comment]
@rubys Quick status update on this PR:
As of this evening,
- every one of the html5lib-tests that doesn't require scripting and that doesn't involve fragment parsing is parsed successfully by nokogumbo, at least when `libxml2`'s headers are available; and
- gumbo's unit tests plus the ones added by @craigbarnes and several I added are all passing. I'm aware of a non-error related bug involving invalid HTML. I added a commented out unit test for that.

There are two issues when the headers aren't available
1. Nokogiri or libxml2 is doing something strange with attributes with the name`xml:lang`. If you look in at [`test/test_tree-construction.rb`](https://github.com/rubys/nokogumbo/pull/84/commits/4b05e33da3f42e9dfbb167c3d7c88508578e4564#diff-e053d760960c00cc074c323565b57e15R155), you can see that I had some trouble detecting that attribute. Hopefully, I can figure out how to work around this.
2. Something is causing `parse` to throw `TypeError: wrong argument type String (expected Data)`. I have no idea what that means, but I should be able to debug.

(Also `&.` doesn't seem to be available on the EOL'd versions of Ruby. I can rewrite it, but I think we should drop those versions anyway since they aren't supported any longer.)

Once these are addressed, I _really_ think we should merge this. The improved testing alone is worth it.

File: ext/nokogumbo/extconf.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 require 'mkmf'
 $CFLAGS += " -std=c99"
 
+CONFIG['warnflags'].gsub!('-Wdeclaration-after-statement', '')
```
Comment (craigbarnes) [inline ext/nokogumbo/extconf.rb]: Can't you get rid of this by just compiling with `-std=c99`?

**Comment by rubys** [comment]
I have no problem with dropping older versions of Ruby on a major release (e.g. 2.0).  I personally don't find `&.` compelling, but ¯\\\_(ツ)_/¯.  Dropping older versions will allow the `Gemfile` to be cleaned up.

I don't have time at the moment to fully review each of the changes, but don't wait on me.  This all looks like good stuff!  Keep up the good work!

Worst case, somebody notices a regression after the release, and has to temporarily pin their application to an older version of nokogumbo until it is resolved.  I'm fine with that.  My personal priorities at the moment is that the code works on macOS High Sierra and Ubuntu 18.04.

**Pull request merged**